Ensure that the hash method does not return null

artonge opened the pull request:
- Follow-up of https://github.com/nextcloud/server/pull/36836
- Fix https://github.com/nextcloud/server/issues/44110

To match:

https://github.com/nextcloud/server/blob/beececf66068f57c416225efcde9b44ce5c2e835/lib/private/Files/View.php#L1050

artonge commented:
/backport to stable29

artonge commented:
/backport to stable28

artonge commented:
/backport to stable27

come-nc commented on lib/private/Files/Storage/Local.php:
```diff
@@ -399,8 +399,8 @@
-	public function hash($type, $path, $raw = false) {
-		return hash_file($type, $this->getSourcePath($path), $raw);
+	public function hash($type, $path, $raw = false): string|false {
+		return hash_file($type, $this->getSourcePath($path), $raw) ?? false;
```
`lib/private/Files/Storage/Wrapper/Availability.php` may return null from `hash()`.
`lib/private/Security/Bruteforce/Backend/MemoryCacheBackend.php` also but I am not sure it’s the same `hash`.

artonge replied:
> lib/private/Files/Storage/Wrapper/Availability.php may return null from hash().

Where do you see that? https://github.com/nextcloud/server/blob/8dcf933cf6a367c1f91531f8eb92153669e5f041/lib/private/Files/Storage/Wrapper/Availability.php#L331-L339

I am also surprised that `hash_file` returns `null` even though the doc says otherwise.

>  lib/private/Security/Bruteforce/Backend/MemoryCacheBackend.php also but I am not sure it’s the same hash.

Indeed, I don't think this is the same `hash`.

come-nc replied:
In the catch branch there is no return

come-nc replied:
Do you have evidence that hash_file returns null apart from this stacktrace? If this is true it should be reported to the PHP project.

artonge committed:
fix(files): Ensure that the hash method does not return null

To match https://github.com/nextcloud/server/blob/beececf66068f57c416225efcde9b44ce5c2e835/lib/private/Files/View.php#L1050

- Fix https://github.com/nextcloud/server/issues/44110

Signed-off-by: Louis Chemineau <louis@chmn.me>

solracsf commented:
> /backport to stable27

Should this be backported to (eol) 27?

come-nc commented on lib/private/Files/Storage/Local.php:
```diff
@@ -399,8 +399,15 @@ public function fopen($path, $mode) {
+		/** @var string|false|null */
+		$hash = hash_file($type, $this->getSourcePath($path), $raw);
+
+		if ($hash === null) {
+			return false;
+		}
```
I read the C code of the PHP function and I really see no possibility for hash_file to return null.
https://github.com/php/php-src/blob/master/ext/hash/hash.c#L460
points to
https://github.com/php/php-src/blob/master/ext/hash/hash.c#L358
It returns false, string or throws.

come-nc replied:
I would suggest to remove this null check, only keep the fix in lib/private/Files/Storage/Wrapper/Availability.php and the strong return type here, and merge that.

susnux commented on lib/private/Files/Storage/Local.php:
```diff
@@ -399,8 +399,15 @@ public function fopen($path, $mode) {
-	public function hash($type, $path, $raw = false) {
-		return hash_file($type, $this->getSourcePath($path), $raw);
+	public function hash($type, $path, $raw = false): string|false {
+		/** @var string|false|null */
+		$hash = hash_file($type, $this->getSourcePath($path), $raw);
+
+		if ($hash === null) {
+			return false;
+		}
+
+		return $hash;
```
```suggestion
		return hash_file($type, $this->getSourcePath($path), $raw);
```

To apply Come's suggestion

skjnldsv committed:
chore: improve hash_file php usage in Local Storage

Co-authored-by: Ferdinand Thiessen <opensource@fthiessen.de>
Signed-off-by: John Molakvoæ <skjnldsv@users.noreply.github.com>
